# Python pandas Practice Problems

Welcome to datascience@berkeley's collection of Python pandas practice problems.


In [2]:
# Import statements

import pandas as pd
import statsmodels.api as sm
import numpy as np

## Importing Data and Making a DataFrame
The statsmodels package (installed in the code cell above) includes built-in datasets. Execute the code below to download data from the [American National Election Studies of 1996](https://www.statsmodels.org/dev/datasets/generated/anes96.html) and print a detailed description of the schema.

The next cell extracts the `Dataset` object from the submodule and saves the `DataFrame` to the variable `df`. In the questions that follow, use `df` when referencing the dataset.

In [3]:
anes96 = sm.datasets.anes96
print(anes96.NOTE)

::

    Number of observations - 944
    Number of variables - 10

    Variables name definitions::

            popul - Census place population in 1000s
            TVnews - Number of times per week that respondent watches TV news.
            PID - Party identification of respondent.
                0 - Strong Democrat
                1 - Weak Democrat
                2 - Independent-Democrat
                3 - Independent-Indpendent
                4 - Independent-Republican
                5 - Weak Republican
                6 - Strong Republican
            age : Age of respondent.
            educ - Education level of respondent
                1 - 1-8 grades
                2 - Some high school
                3 - High school graduate
                4 - Some college
                5 - College degree
                6 - Master's degree
                7 - PhD
            income - Income of household
                1  - None or less than $2,999
                2  - $3,000-$4,9

In [4]:
dataset_anes96 = anes96.load_pandas()
df = dataset_anes96.data

**In this notebook, UC Berkeley has provided the code to download the dataset and the questions to answer.** 
**The dataset has been assigned to the variable df.**

## 1. DataFrame Basic Properties Exercise

Our DataFrame (`df`) contains data on registered voters in the United States, including demographic information and political preference. Using `pandas`, print the first 5 rows of the DataFrame to get a sense of what the data looks like. Next, answer the following questions:


*   How many observations are in the DataFrame?
*   How many variables are measured (how many columns)?
*   What is the age of the youngest person in the data? The oldest?
*   How many days a week does the average respondent watch TV news (round to the nearest tenth)?
*   Check for missing values. Are there any?






In [5]:
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  educ  income  vote  \
0    0.0     7.0     7.0     1.0     6.0  6.0  36.0   3.0     1.0   1.0   
1  190.0     1.0     3.0     3.0     5.0  1.0  20.0   4.0     1.0   0.0   
2   31.0     7.0     2.0     2.0     6.0  1.0  24.0   6.0     1.0   0.0   
3   83.0     4.0     3.0     4.0     5.0  1.0  28.0   6.0     1.0   0.0   
4  640.0     7.0     5.0     6.0     4.0  0.0  68.0   6.0     1.0   0.0   

   logpopul  
0 -2.302585  
1  5.247550  
2  3.437208  
3  4.420045  
4  6.461624

**The head method gives us the first five rows in the DataFrame.**

In [6]:
df.shape

(944, 11)

**The shape attribute allows us to see the dimensions of the DataFrame.
df has 944 rows and 11 columns.**

In [7]:
df['age'].max()

91.0

**First, we select the age column in df. Then, we use the max function to find and return the maximum value.**

In [8]:
df['age'].min()

19.0

**Similar to the max function, the opposing min function returns the minimum value from the age column.**

In [9]:
round(df['TVnews'].mean(), 1)

3.7

**By using the round function, we can take the average of the TVnews column, and round it to one decimal place**

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   popul     944 non-null    float64
 1   TVnews    944 non-null    float64
 2   selfLR    944 non-null    float64
 3   ClinLR    944 non-null    float64
 4   DoleLR    944 non-null    float64
 5   PID       944 non-null    float64
 6   age       944 non-null    float64
 7   educ      944 non-null    float64
 8   income    944 non-null    float64
 9   vote      944 non-null    float64
 10  logpopul  944 non-null    float64
dtypes: float64(11)
memory usage: 81.2 KB


**The info method gives us a breakdown of the columns. Here, we see that there are 944 total entries and for all the columns, there are no null values, making it reliable to use this data. We also see that all columns are floats.**

## 2. Data Processing Exercise

We want to adjust the dataset for our use. Do the following:


*   Rename the `educ` column `education`.
*   Create a new column called `party` based on each respondent's answer to `PID`. `party` should equal `Democrat` if the respondent selected either Strong Democrat or Weak Democrat. `party` will equal `Republican` if the respondent selected Strong or Weak Republican for `PID` and `Independent` if they selected anything else.
*   Create a new column called `age_group` that buckets respondents into the following categories based on their `age`: 18-24, 25-34, 35-44, 45-54, 55-64, and 65 and over.



In [11]:
df = df.rename(columns= {'educ':'education'})
df

popul  TVnews  selfLR  ClinLR  DoleLR  PID   age  education  income  \
0      0.0     7.0     7.0     1.0     6.0  6.0  36.0        3.0     1.0   
1    190.0     1.0     3.0     3.0     5.0  1.0  20.0        4.0     1.0   
2     31.0     7.0     2.0     2.0     6.0  1.0  24.0        6.0     1.0   
3     83.0     4.0     3.0     4.0     5.0  1.0  28.0        6.0     1.0   
4    640.0     7.0     5.0     6.0     4.0  0.0  68.0        6.0     1.0   
..     ...     ...     ...     ...     ...  ...   ...        ...     ...   
939    0.0     7.0     7.0     1.0     6.0  4.0  73.0        6.0    24.0   
940    0.0     7.0     5.0     2.0     6.0  6.0  50.0        6.0    24.0   
941    0.0     3.0     6.0     2.0     7.0  5.0  43.0        6.0    24.0   
942    0.0     6.0     6.0     2.0     5.0  6.0  46.0        7.0    24.0   
943   18.0     7.0     4.0     2.0     6.0  3.0  61.0        7.0    24.0   

     vote  logpopul  
0     1.0 -2.302585  
1     0.0  5.247550  
2     0.0  3.437208  
3     0.0  4.420045  
4     0.0  6.461624  
..    ...       ...  
939   1.0 -2.302585  
940   1.0 -2.302585  
941   1.0 -2.302585  
942   1.0 -2.302585  
943   1.0  2.895912  

[944 rows x 11 columns]

 **Here, we use the rename function, specifically selecting the 'educ' column
 and replacing it with 'education'**

In [12]:
df.insert(loc = 6, column = 'party', value = np.where(df['PID']<=1, 'Democrat', np.where(df['PID']<=4, 'Independent', 'Republican')))

**Here, we use the insert statement so we can utilize the loc parameter to choose the index and place the column next to the PID for easier viewing. Then, we use a series of np.where functions to account for the different parties.**

In [13]:
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID       party   age  education  \
0    0.0     7.0     7.0     1.0     6.0  6.0  Republican  36.0        3.0   
1  190.0     1.0     3.0     3.0     5.0  1.0    Democrat  20.0        4.0   
2   31.0     7.0     2.0     2.0     6.0  1.0    Democrat  24.0        6.0   
3   83.0     4.0     3.0     4.0     5.0  1.0    Democrat  28.0        6.0   
4  640.0     7.0     5.0     6.0     4.0  0.0    Democrat  68.0        6.0   

   income  vote  logpopul  
0     1.0   1.0 -2.302585  
1     1.0   0.0  5.247550  
2     1.0   0.0  3.437208  
3     1.0   0.0  4.420045  
4     1.0   0.0  6.461624

In [14]:
df.insert(loc = 8, column = 'age_group', value = np.where(df['age']<=24, '18-24', np.where(df['age']<=34, '24-34', np.where(df['age']<=44, '35-44', np.where(df['age']<=54, '45-54', np.where(df['age']<=64, '55-64', '65+'))))))

**Here, we use the insert statement to add the 'age_group' column again and choose the index after the 'age' column. Then, we use a series of np.where functions to account for the different age groups.**

In [15]:
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID       party   age age_group  \
0    0.0     7.0     7.0     1.0     6.0  6.0  Republican  36.0     35-44   
1  190.0     1.0     3.0     3.0     5.0  1.0    Democrat  20.0     18-24   
2   31.0     7.0     2.0     2.0     6.0  1.0    Democrat  24.0     18-24   
3   83.0     4.0     3.0     4.0     5.0  1.0    Democrat  28.0     24-34   
4  640.0     7.0     5.0     6.0     4.0  0.0    Democrat  68.0       65+   

   education  income  vote  logpopul  
0        3.0     1.0   1.0 -2.302585  
1        4.0     1.0   0.0  5.247550  
2        6.0     1.0   0.0  3.437208  
3        6.0     1.0   0.0  4.420045  
4        6.0     1.0   0.0  6.461624

## 3. Filtering Data Exercise

Use the filtering method to find all the respondents who have the impression that Bill Clinton is moderate or conservative (`ClinLR` equals 4 or higher). How many respondents are in this subset?

Among these respondents, how many have a household income less than $50,000 and attended at least some college?

In [16]:
clint = df[df['ClinLR'] >= 4]
clint.shape

(282, 13)

**Here, we define clint as the respondents whose ClinLR are greater than 4. By looking at the shape, we see that there are 282 respondents who believe Bill Clinton is moderate/conservative.**

In [17]:
clint[(clint['income']<20) & (clint['education'] >3)].shape

(98, 13)

**By further filtering the clint dataframe for income and education, we see that 98 respondents, or about 35% of the respondents who believe Bill Clinton is moderate or conservative, have a household income of less than $50,000 and attended some college or more.**

## 4. Calculating From Data Exercise

For each of the below match-ups, choose the group that is more likely to vote for Bill Clinton. You can calculate this using the percentage of each group that intends to vote for Clinton (`vote`). Which match-up was the closest? Which had the biggest difference?

Another way to think about this: Given that a respondent is a Democrat, there is a ____ percent chance they will vote for Clinton. How does this value change if the respondent is a Republican?

*   Democrats or Republicans
*   People younger than 44 or People 44 and older
*   People who watch TV news at least 6 days a week or People who watch TV news less than 3 days a week

In [18]:
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID       party   age age_group  \
0    0.0     7.0     7.0     1.0     6.0  6.0  Republican  36.0     35-44   
1  190.0     1.0     3.0     3.0     5.0  1.0    Democrat  20.0     18-24   
2   31.0     7.0     2.0     2.0     6.0  1.0    Democrat  24.0     18-24   
3   83.0     4.0     3.0     4.0     5.0  1.0    Democrat  28.0     24-34   
4  640.0     7.0     5.0     6.0     4.0  0.0    Democrat  68.0       65+   

   education  income  vote  logpopul  
0        3.0     1.0   1.0 -2.302585  
1        4.0     1.0   0.0  5.247550  
2        6.0     1.0   0.0  3.437208  
3        6.0     1.0   0.0  4.420045  
4        6.0     1.0   0.0  6.461624

In [34]:
parties = df.groupby(['party','vote']).size().to_frame('num_of_resp').reset_index()
num_of_dems = df[df['party'] == 'Democrat'].shape[0]
num_of_rep = df[df['party'] == 'Republican'].shape[0]
def perc(df):
    if df['party'] == 'Democrat':
        return round(df['num_of_resp']/num_of_dems*100, 2)
    elif df['party'] == 'Republican':
        return round(df['num_of_resp']/num_of_rep*100, 2)
parties['percentage'] = parties.apply(perc, axis = 1)
parties[((parties['party'] == 'Democrat') | (parties['party'] == 'Republican')) & (parties['vote']==0)]

party  vote  num_of_resp  percentage
0    Democrat   0.0          366       96.32
4  Republican   0.0           34       10.46

**Here, we want to group by the party and the person the respondents are voting for. Then, we can use a function to find the percent of each party that is voting for Clinton.**

**We see that 96.32% of Democrats expect to vote for Clinton, while only 10.46% of Republicans will.**

In [20]:
def age_checker(df):
    if df['age'] <44:
        return '<44'
    else:
        return '>=44'
df['age_bucket'] = df.apply(age_checker, axis = 1)
ages = df.groupby(['age_bucket', 'vote']).size().to_frame('num_of_resp').reset_index()
young = df[df['age'] < 44].shape[0]
old = df[df['age'] >= 44].shape[0]
def percent(df):
    if df['age_bucket'] =='<44':
        return round(df['num_of_resp']/young*100, 2)
    if df['age_bucket'] == '>=44':
        return round(df['num_of_resp']/old*100, 2)
ages['percent'] = ages.apply(percent, axis=1)
ages[ages['vote'] == 0]

age_bucket  vote  num_of_resp  percent
0        <44   0.0          276    59.48
2       >=44   0.0          275    57.29

**Here, we use a similar logic to the one above. We add an age_checker function to categorize each respondent as either younger or older than 44.Then group the data by those two categories.** 

**Here, we see that regardless of the age, approximately the same percent of respondents expect to vote for Clinton. However, those who are younger than 44 are slightly more likely to vote for Clinton.**

In [21]:
def day_checker(df):
    if df['TVnews']<3:
        return '<3'
    elif df['TVnews']>5:
        return '>5'
    else:
        return '3-5'
df['days_watch'] = df.apply(day_checker, axis = 1)
days = df.groupby(['days_watch', 'vote']).size().to_frame('num_of_resp').reset_index()
num_of_low = df[df['days_watch'] == '<3'].shape[0]
num_of_hi = df[df['days_watch'] == '>5'].shape[0]
def perc(df):
    if df['days_watch'] == '<3':
        return round(df['num_of_resp']/num_of_low*100, 2)
    elif df['days_watch'] =='>5':
        return round(df['num_of_resp']/num_of_hi*100, 2)
days['percent'] = days.apply(perc, axis = 1)
days[(days['vote']==0) & (days['days_watch'] != '3-5')]

days_watch  vote  num_of_resp  percent
2         <3   0.0          207    55.50
4         >5   0.0          185    57.81

**Here, we use a similar logic, but accounting for the number of days each respondent watches the news.**

**We see that, similar to above, there is not a significant difference in the percent of respondents that expect to vote for Clinton regardless of if they watch the news fewer or more days per week. Although, those that watch the news more often per week are more likely to vote for Clinton.**

## 5. Grouping Data Exercise

Use the `groupby()` method to bucket respondents by `age_group`. Which age group is the most conservative? Which watches TV news the least?

Next, calculate 5 percentile groups based on income. Group the dataset by these percentiles. Which income bracket is the most liberal? Which is the most conservative? The oldest? Highest educated?

In [22]:
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID       party   age age_group  \
0    0.0     7.0     7.0     1.0     6.0  6.0  Republican  36.0     35-44   
1  190.0     1.0     3.0     3.0     5.0  1.0    Democrat  20.0     18-24   
2   31.0     7.0     2.0     2.0     6.0  1.0    Democrat  24.0     18-24   
3   83.0     4.0     3.0     4.0     5.0  1.0    Democrat  28.0     24-34   
4  640.0     7.0     5.0     6.0     4.0  0.0    Democrat  68.0       65+   

   education  income  vote  logpopul age_bucket days_watch  
0        3.0     1.0   1.0 -2.302585        <44         >5  
1        4.0     1.0   0.0  5.247550        <44         <3  
2        6.0     1.0   0.0  3.437208        <44         >5  
3        6.0     1.0   0.0  4.420045        <44        3-5  
4        6.0     1.0   0.0  6.461624       >=44         >5

In [31]:
df.groupby('age_group')['selfLR'].mean().to_frame('average_selfLR').sort_values('average_selfLR', ascending = False).reset_index()

age_group  average_selfLR
0     55-64        4.532258
1       65+        4.529412
2     45-54        4.285714
3     35-44        4.257143
4     24-34        4.217391
5     18-24        4.000000

**Here, the closer the selfLR is to 7, the more conservative. Therefore, the 55-64 age group is the most conservative. It is also interesting to note that among those younger than 54, as the age brackets get younger, the average political view becomes slightly more liberal.**

In [32]:
df.groupby('age_group')['TVnews'].mean().to_frame('average_TVnews').sort_values('average_TVnews').reset_index()

age_group  average_TVnews
0     18-24        2.358491
1     24-34        2.570652
2     35-44        3.118367
3     45-54        3.916667
4     55-64        4.516129
5       65+        5.523529

**Here, TVnews represents the number of times the respondent watches TV in a week. Therefore, the 18-24 age group watch the TV the least amount of times per week.** 

**We see a trend: as the age groups get older, the average number of days the respondents watch TV news per week increases. This could be a reflection of the growing criticism towards fake news and biased news sources/outlets among the younger generation.**

## 6. Voting Across the Aisle

We are interested in learning more about respondents who's political views differ strongly from the candidate they expect to vote for. Using `selfLR`, `vote`, `ClinLR`, and `DoleLR`, work through the following questions. Your interpretation may differ from the answer key.

*   What is the largest recorded difference between a respondent's political leaning and their impression of their intended candidate's political leaning?
*   How many respondents exhibit a difference of that magnitude?
*   Make a separate DataFrame called `sway` that only includes voters who exhibit a difference greater than |3|.
*   Among those in `sway`, are respondents more likely to be voting for a candidate more conservative or more liberal than their own political leaning?
*   In `sway`, which candidate is the more popular choice?



In [25]:
df.head()

popul  TVnews  selfLR  ClinLR  DoleLR  PID       party   age age_group  \
0    0.0     7.0     7.0     1.0     6.0  6.0  Republican  36.0     35-44   
1  190.0     1.0     3.0     3.0     5.0  1.0    Democrat  20.0     18-24   
2   31.0     7.0     2.0     2.0     6.0  1.0    Democrat  24.0     18-24   
3   83.0     4.0     3.0     4.0     5.0  1.0    Democrat  28.0     24-34   
4  640.0     7.0     5.0     6.0     4.0  0.0    Democrat  68.0       65+   

   education  income  vote  logpopul age_bucket days_watch  
0        3.0     1.0   1.0 -2.302585        <44         >5  
1        4.0     1.0   0.0  5.247550        <44         <3  
2        6.0     1.0   0.0  3.437208        <44         >5  
3        6.0     1.0   0.0  4.420045        <44        3-5  
4        6.0     1.0   0.0  6.461624       >=44         >5

In [26]:
def diff(df):
    if df['vote'] == 0:
        return df['selfLR']-df['ClinLR']
    else:
        return df['selfLR']-df['DoleLR']
df['difference'] = df.apply(diff, axis = 1)
df['difference'].abs().max()

6.0

**Here, we can use a function to calculate the difference in the selfLR and the ClinLR/DoleLR, depending on who the respondent is expected to vote for.** 

**We see that the maximum difference between the self reported political leaning and that of the candidate they expect to vote for is 6.**

In [27]:
df[df['difference'] == df['difference'].abs().max()].shape[0]

1

**There is only one respondent whose difference is the maximum difference.**

In [36]:
sway = df.copy()[df['difference'].abs() >= 3]
def diff_check(df):
    if df['difference']<0:
        return 'self_more_lib'
    else:
        return 'self_more_con'
sway['pol_view'] = sway.apply(diff_check, axis=1)
sway.groupby('pol_view').size().to_frame('num_of_resp').reset_index()

pol_view  num_of_resp
0  self_more_con           40
1  self_more_lib           16

**For reference, if the difference is positive, the respondent is more conservative than the candidate. If the difference is negative, the respondent is more liberal than the candidate.**

**Among those who have a difference of 3 or greater, 40 respondents view themselves as more conservative than the candidate they expect to vote for and 16 respondents view themselves as more liberal than the candidate.** 

**Therefore, a majority of respondents are likely to vote for the the candidate that is more liberal than themselves.**

In [29]:
sway.groupby('vote').size().to_frame('num_of_resp').reset_index()

vote  num_of_resp
0   0.0           48
1   1.0            8

**Here, we see that majority of the respondents that have a political leaning significantly more different to their candidate are more likely to vote for Clint than Dole.** 

# BSD 3-Clause License

*Copyright (c) 2022, UC Berkeley School of Information*

*All rights reserved.*

*Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:*

*1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.*

*2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.*

*3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.*

*THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.*